src/Streamly/Internal/Data/Stream/IsStream/Top.hs

```diff
@@ -0,0 +1,571 @@
+import qualified Streamly.Internal.Data.Stream.IsStream.Transform as Stream
+import qualified Streamly.Internal.Data.Stream.StreamK as StreamK
+
+import Prelude hiding (filter, zipWith, concatMap, concat)
+
+------------------------------------------------------------------------------
+-- Sampling
+------------------------------------------------------------------------------
+
+-- | @sampleFromthen offset stride@ samples the element at @offset@ index and
+-- then every element at strides of @stride@.
+--
+-- @
+-- >>> Stream.toList $ Stream.sampleFromThen 3 2 $ S.enumerateFromTo 0 10
+-- > [1,4,7]
+-- @
+--
+-- /Internal/
+--
+{-# INLINE sampleFromThen #-}
+sampleFromThen :: (IsStream t, Monad m, Functor (t m)) =>
+    Int -> Int -> t m a -> t m a
+sampleFromThen size offset =
+    Stream.with Stream.indexed Stream.filter
+        (\(i, _) -> (i + offset) `mod` size == 0)
+
+-- | Continuously evaluate the input stream and sample the last event in time
+-- window of @n@ seconds.
+--
+-- This is also known as @throttle@ in some libraries.
+--
+-- @sampleIntervalEnd n = Stream.intervalsOf n Fold.last@
+--
+-- /Internal/
+--
+{-# INLINE sampleIntervalEnd #-}
+sampleIntervalEnd :: (IsStream t, MonadAsync m, Functor (t m)) =>
+    Double -> t m a -> t m a
+sampleIntervalEnd n = Stream.catMaybes . Stream.intervalsOf n Fold.last
+
+-- | Like 'sampleInterval' but samples at the beginning of the time window.
+--
+-- @sampleIntervalStart n = Stream.intervalsOf n Fold.head@
+--
+-- /Internal/
+--
+{-# INLINE sampleIntervalStart #-}
+sampleIntervalStart :: (IsStream t, MonadAsync m, Functor (t m)) =>
+    Double -> t m a -> t m a
+sampleIntervalStart n = Stream.catMaybes . Stream.intervalsOf n Fold.head
+
+-- | Sample one event at the end of each burst of events.  A burst is a group
+-- of events close together in time, it ends when an event is spaced by more
+-- than the specified time interval from the previous event.
+--
+-- This is known as @debounce@ in some libraries.
+--
+-- The clock granularity is 10 ms.
+--
+-- /Internal/
+--
+{-# INLINE sampleBurstEnd #-}
+sampleBurstEnd :: (IsStream t, MonadAsync m, Functor (t m)) =>
+    Double -> t m a -> t m a
+sampleBurstEnd gap =
+    let f (t1, _) (t2, _) =
+            t2 - t1 >= toRelTime64 (NanoSecond64 (round (gap * 10^(9::Int))))
+    in fmap snd
+        . Stream.catMaybes
+        . Stream.groupsByRolling f Fold.last
+        . Stream.timeIndexed
+
+-- | Like 'sampleBurstEnd' but samples the event at the beginning of the burst
+-- instead of at the end of it.
+--
+-- /Internal/
+--
+{-# INLINE sampleBurstStart #-}
+sampleBurstStart :: (IsStream t, MonadAsync m, Functor (t m)) =>
+    Double -> t m a -> t m a
+sampleBurstStart gap =
+    let f (t1, _) (t2, _) =
+            t2 - t1 >= toRelTime64 (NanoSecond64 (round (gap * 10^(9::Int))))
+    in fmap snd
+        . Stream.catMaybes
+        . Stream.groupsByRolling f Fold.last
```

Comment:
```suggestion
        . Stream.groupsByRolling f Fold.head
```
I did not verify the behavior. Felt symmetric.

Reply:
Yes, that's a (typo) bug.